File: src/main/java/ar/com/florius/aao/TagInterceptor.java
package ar.com.florius.aao;

import ar.com.florius.aao.semilattice.TagName;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.implementation.bind.annotation.AllArguments;
import net.bytebuddy.implementation.bind.annotation.Origin;
import net.bytebuddy.implementation.bind.annotation.RuntimeType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static ar.com.florius.aao.Tag.tag;

public class TagInterceptor<T> {
    final Logger logger = LoggerFactory.getLogger(Tag.class);
    private final SafeTag<T> safeTag;

    public TagInterceptor(SafeTag<T> safeTag) {
        this.safeTag = safeTag;
    }

    @RuntimeType
    @SuppressWarnings("unused")
    public Object intercept(@Origin Method method, @AllArguments Object[] args) throws Exception {
        switch (method.getName()) {
            case "getTag":
                return this.safeTag.getTag();
            case "getValue":
                return this.safeTag.getValue();
            default:
                return dispatch(method, args);
        }
    }

    private Object dispatch(Method method, Object[] args) throws IllegalAccessException, InvocationTargetException {
        List<TagName> argsTag = Arrays.stream(args)
                .map(Tag::safeToTaggable)
                .map(objectTaggable -> objectTaggable.map(Taggable::getTag).orElse(TagName.NoTag.INSTANCE))
                .collect(Collectors.toList());

        TagName newTag = this.safeTag.operate(argsTag);
        Object result = method.invoke(this.safeTag.getValue(), args);
        TypeDescription resultType = TypeDescription.ForLoadedType.of(result.getClass());

        if (resultType.isPrimitive() || resultType.isArray() || resultType.isFinal()) {
            logger.warn("Cannot tag primitive, array or final types ({}). Returning untagged ({})", resultType, result);
            return result;
        } else {
            return tag(result, newTag);
        }
    }

}

File: src/main/kotlin/ar/com/florius/aao/TagInterceptor.kt
package ar.com.florius.aao

import ar.com.florius.aao.Tag.tag
import ar.com.florius.aao.semilattice.Namespace
import ar.com.florius.aao.strict.Tag.safeToTaggable
import net.bytebuddy.description.type.TypeDescription
import net.bytebuddy.implementation.bind.annotation.AllArguments
import net.bytebuddy.implementation.bind.annotation.Origin
import net.bytebuddy.implementation.bind.annotation.RuntimeType
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

class TagInterceptor<T>(private val safeTag: SafeTag<T>) {
    private val logger = LoggerFactory.getLogger(TagInterceptor::class.java)

    @RuntimeType
    @Suppress("UNUSED")
    fun intercept(@Origin method: Method, @AllArguments args: Array<Any?>): Any? {
        return when (method.name) {
            "getTag" -> safeTag.tag
            "getValue" -> safeTag.value
            else -> dispatch(method, args)
        }
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    private fun dispatch(method: Method, args: Array<Any?>): Any? {
        val argsTag: List<Taggable<Any?, Namespace>> = args
            .map { obj -> safeToTaggable<Any?, Namespace>(obj).orElse(SafeTag(obj, Namespace.min)) }

        val result = safeTag.apply(method, argsTag)
        if (result.tag == Namespace.Incompatible) {
            throw IncompatibleTagsException(result, argsTag)
        }

        val resultType = result.value?.let { TypeDescription.ForLoadedType.of(it.javaClass) }

        if (resultType == null || resultType.isPrimitive || resultType.isArray || resultType.isFinal) {
            logger.warn(
                "Cannot tag null, primitive, array or final types ({}). Returning untagged ({})",
                resultType,
                result
            )
            return result.value
        }
        return tag(result.value, result.tag)
    }
}